### PR TITLE
[CI] Fix unit test TestConfigurationMgr random failure   

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  Examples:
* The CI randomly failed with the following error:
[1630514648.709927][29145:29145] CHIP:DL: writing settings to file (/tmp/chip_factory.ini.tmp)
[1630514648.709969][29145:29145] CHIP:DL: failed to rename (/tmp/chip_factory.ini.tmp), No such file or directory (2)
[1630514648.709980][29145:29145] CHIP:DL: Configuration Manager initialization failed: ../../src/platform/Linux/CHIPLinuxStorageIni.cpp:113: CHIP Error 0x000000AF: Write to file failed

The root cause of this failure is the unit test run in parallel in CI and all share the same temp file on the storage which cause race condition.

#### Change overview
Use ramdam temporary file name to prevent racing on the shared file.

#### Testing
How was this tested? (at least one bullet point required)
* Covered by CI

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Comment by andy31415** [comment]
Will this work for chip-tool? I believe chip-tool assumes persistence. We may see this in the 'TESTS' where we have step(1) running pairing and step(2) running some test.

**Review by andy31415 — APPROVED** [review]

**Comment by yufengwangca** [comment]
> Will this work for chip-tool? I believe chip-tool assumes persistence. We may see this in the 'TESTS' where we have step(1) running pairing and step(2) running some test.

Yes, it seems the race only occurs on ephemeral temp files, no change for the final configuration files

**Review by mspang — CHANGES_REQUESTED** [review]
Please don't use pid to create temporary files. It is

- Not sufficiently unique
- Not atomic

This problem probably cannot be solved with ofstream. The solution is to use an API for temporary files (at least name generation). The file _must_ be opened with `O_EXCL|O_CREAT` to avoid the race.

**Comment by mspang** [comment]
> Please don't use pid to create temporary files. It is
> 
> * Not sufficiently unique
> * Not atomic
> 
> This problem probably cannot be solved with ofstream. The solution is to use an API for temporary files (at least name generation). The file _must_ be opened with `O_EXCL|O_CREAT` to avoid the race.

If we're sure we've created the file it should be safe to open again with ofstream.

**Comment by yufengwangca** [comment]
> > Please don't use pid to create temporary files. It is
> > 
> > * Not sufficiently unique
> > * Not atomic
> > 
> > This problem probably cannot be solved with ofstream. The solution is to use an API for temporary files (at least name generation). The file _must_ be opened with `O_EXCL|O_CREAT` to avoid the race.
> 
> If we're sure we've created the file it should be safe to open again with ofstream.

The problem is the file is deleted by another instance of program which run into ofs.close(); The race is not among multiple threads, but multiple programs which use the same tmp file name.

**Comment by mspang** [comment]
> > > Please don't use pid to create temporary files. It is
> > > 
> > > * Not sufficiently unique
> > > * Not atomic
> > > 
> > > This problem probably cannot be solved with ofstream. The solution is to use an API for temporary files (at least name generation). The file _must_ be opened with `O_EXCL|O_CREAT` to avoid the race.
> > 
> > 
> > If we're sure we've created the file it should be safe to open again with ofstream.
> 
> The problem is the file is deleted by another instance of program which run into ofs.close(); The race is not among multiple threads, but multiple programs which use the same tmp file name.

I assume you mean that the race is with rename(). Closing doesn't delete the file.

My position is that if we are going to try to work around this problem by using a unique temporary file, at least the creation of the unique temporary file should be done in the correct way. How to do it is well known and available in the standard library as mkstemp is for (or mkdtemp would work too). Doing it the wrong way is either still subject to races or is a vector for symlink attacks. There is no point in doing it a worse way when the standard library provides the implementation.

Using a unique name still does not solve the general problem that concurrent unit tests behavior might be affected by shared storage (if they are, it will most likely be in a subtle and nondeterministic way due to the concurrency, just like we have with this race).

**Comment by mspang** [comment]
> > > Please don't use pid to create temporary files. It is
> > > 
> > > * Not sufficiently unique
> > > * Not atomic
> > > 
> > > This problem probably cannot be solved with ofstream. The solution is to use an API for temporary files (at least name generation). The file _must_ be opened with `O_EXCL|O_CREAT` to avoid the race.
> > 
> > 
> > If we're sure we've created the file it should be safe to open again with ofstream.
> 
> Could you explain more? If there are two instance of programs try to open the same temp file with both O_CREAT and O_EXCL are set, then the second open should fail since the specified file already exists. The CI still fails in this case.

The answer is that if open with O_EXCL|O_CREAT fails, you must retry with a different name.

**Review by mspang — CHANGES_REQUESTED** [review]
Please don't const_cast a const string argument.

File: src/platform/Linux/CHIPLinuxStorageIni.cpp [hunk 1]
```diff
@@ -88,20 +88,20 @@ CHIP_ERROR ChipLinuxStorageIni::AddConfig(const std::string & configFile)
...
-    std::string tmpPath = configFile;
-
-    tmpPath.append(".tmp");
-
-    ofs.open(tmpPath, std::ofstream::out | std::ofstream::trunc);
+    CHIP_ERROR retval   = CHIP_NO_ERROR;
+    std::string tmpPath = configFile + "-XXXXXX";
 
-    if (ofs.is_open())
+    int fd = mkstemp(const_cast<char *>(tmpPath.c_str()));
```
Comment (mspang) [inline src/platform/Linux/CHIPLinuxStorageIni.cpp:94]: mkstemp(&tmpPath[0]) ?

**Comment by mspang** [comment]
> Please don't const_cast a const string argument.

Ah sorry I see the copy. I'd still avoid it though.

**Review by mspang — APPROVED** [review]

File: src/platform/Linux/CHIPLinuxStorageIni.cpp (hunk 1 above)
Comment (yufengwangca) [reply src/platform/Linux/CHIPLinuxStorageIni.cpp]: Acked

**Comment by andy31415** [comment]
@chrisdecenzo @jelderton @jmartinez-silabs @LuDuda @saurabhst ?

**Review by LuDuda — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Pull request merged**